private/requestcompression/request_compression.go

```diff
@@ -0,0 +1,105 @@
+package requestcompression
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"github.com/aws/smithy-go/middleware"
+	"github.com/aws/smithy-go/transport/http"
+	"io"
+)
+
+// AddRequestCompression add requestCompression middleware to op stack
+func AddRequestCompression(stack *middleware.Stack, DisableRequestCompression bool, RequestMinCompressSizeBytes int64) error {
+	return stack.Build.Add(&requestCompression{
```

Comment:
"Clients MUST encode the request content prior to request signing, otherwise the request signature will mismatch."

According to SEP requirement, I agree with you that this middleware should be injected at an earlier stage